sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -5970,35 +6032,149 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
+  // Indicator for whether batching is allowed. This may be changed later in
+  // this function, but allow it by default.
   bool OkToBatch = true;
-  pi_command_list_ptr_t CommandList{};
-  if (auto Res = Queue->Context->getAvailableCommandList(
-          Queue, CommandList, UseCopyEngine, OkToBatch))
-    return Res;
 
-  ze_event_handle_t ZeEvent = nullptr;
-  auto Res = createEventAndAssociateQueue(Queue, Event, PI_COMMAND_TYPE_USER,
-                                          CommandList);
-  if (Res != PI_SUCCESS)
-    return Res;
-  ZeEvent = (*Event)->ZeEvent;
-  (*Event)->WaitList = TmpWaitList;
+  // If we have a list of events to make the barrier from, then we can create a
+  // barrier on these and use the resulting event as our future barrier.
+  if (NumEventsInWaitList) {
+    // Get an arbitrary command-list in the queue.
+    pi_command_list_ptr_t CmdList;
+    if (auto Res = Queue->Context->getAvailableCommandList(
+            Queue, CmdList,
+            /*UseCopyEngine=*/false, OkToBatch))
+      return Res;
 
-  ZE_CALL(zeCommandListAppendBarrier,
-          (CommandList->first, ZeEvent, (*Event)->WaitList.Length,
-           (*Event)->WaitList.ZeEventList));
+    // Retain the events as they will be owned by the result event.
+    _pi_ze_event_list_t TmpWaitList;
+    if (auto Res = TmpWaitList.createAndRetainPiZeEventList(
+            NumEventsInWaitList, EventWaitList, Queue,
+            CmdList->second.isCopy(Queue)))
+      return Res;
 
-  // Execute command list asynchronously as the event will be used
-  // to track down its completion.
-  return Queue->executeCommandList(CommandList, false, OkToBatch);
+    // Insert the barrier into the command-list and execute.
+    if (auto Res = insertBarrierIntoCmdList(CmdList, TmpWaitList, *Event))
+      return Res;
+    if (auto Res = Queue->executeCommandList(CmdList, false, OkToBatch))
+      return Res;
+
+    // Retain and save the resulting event for future commands.
+    PI_CALL(piEventRetain(*Event));
+    Queue->ActiveBarriers.push_back(*Event);
+    return PI_SUCCESS;
+  }
+
+  // Since there are no events to explicitly create a barrier for, we are
+  // inserting a queue-wide barrier. As such, the barrier will also encapsulate
+  // the active barriers, so we can release and clear the active barriers list.
+  // Doing it early prevents potential additional barriers from implicitly being
+  // appended.
+  for (pi_event &E : Queue->ActiveBarriers)
+    PI_CALL(piEventRelease(E));
+  Queue->ActiveBarriers.clear();
+
+  // Get command lists for each command queue.
+  std::vector<pi_command_list_ptr_t> CmdLists;
+  if (UseImmediateCommandLists) {
+    // If immediate command lists are being used, each will act as their own
+    // queue, so we must insert a barrier into each.
+    CmdLists.reserve(Queue->CommandListMap.size());
+    for (auto It = Queue->CommandListMap.begin();
+         It != Queue->CommandListMap.end(); ++It)
+      CmdLists.push_back(It);
```

Comment:
we are going to create multiple immediate command-lists too for a single SYCL queue (returned by getImmCmdList) so maybe this should already account for this to avoid hard to find debugging later?

Reply:
"Active" barriers are inserted into any immediate command-list gotten through `getImmCmdList` currently (in `getAvailableCommandList`). Here the most important aspect is that we get command lists for all queues that potentially have had previous work enqueued, which in this case mean all existing immediate command lists.